### ISSUE TITLE
Unable to get local issuer certificate error for Generate Release Notes (Crossplatform)

### ISSUE THREAD

**Issue by mayeager22** [rfennell/AzurePipelines#801]
I have the "Generate Release Notes (Crossplatform)" extension installed and am trying to use version 3.* in a release pipeline task in Azure DevOps Server 2019.

The task succeeded, but no file is generated and the following stack trace is in the log file:

**### **2020-09-01T20:04:42.2521115Z ##[section]Starting: Generate Release Notes**
2020-09-01T20:04:42.3806890Z ==============================================================================
2020-09-01T20:04:42.3807064Z Task         : Generate Release Notes (Crossplatform)
2020-09-01T20:04:42.3807108Z Description  : Generates a release notes file in a format of your choice from the build or release history
2020-09-01T20:04:42.3807148Z Version      : 3.8.3
2020-09-01T20:04:42.3807197Z Author       : Black Marble
2020-09-01T20:04:42.3807652Z Help         : Version: 3.8.3. [More Information](https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task/)
2020-09-01T20:04:42.3807688Z ==============================================================================
2020-09-01T20:04:43.8123505Z (node:8972) UnhandledPromiseRejectionWarning: Error: unable to get local issuer certificate
2020-09-01T20:04:43.8251101Z     at TLSSocket.onConnectSecure (_tls_wrap.js:1049:34)
2020-09-01T20:04:43.8251694Z     at TLSSocket.emit (events.js:182:13)
2020-09-01T20:04:43.8252229Z     at TLSSocket._finishInit (_tls_wrap.js:631:8)
2020-09-01T20:04:43.8252898Z (node:8972) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
2020-09-01T20:04:43.8253201Z (node:8972) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
2020-09-01T20:04:43.8319478Z ##[section]Finishing: Generate Release Notes**

**Comment by rfennell** [comment]
I will have to build an Azure DevOps Server 2019 to test this scenario.

If might help if you run a build the the variable `system.debug=true` this will give more information and show us how far it is getting.

**Comment by rfennell** [comment]
This seems similar to this [Azure Pipelines issue](https://github.com/microsoft/azure-pipelines-tasks/issues/11508).

So what is the version of Node installed on the Agent? 

- You could install the latest version, at least the LTS version https://nodejs.org/en/download/ as this task requires Node 10
- or can use [Node.js Tool Installer task](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/node-js?view=azure-devops) to install the latest version of the node.js in your agent as the first step in your pipelines

**Comment by rfennell** [comment]
I have just tested this on an Azure DevOps Server 2019.Update1.1 instance and it has worked. Note that I did have to install the [current version of Node](https://nodejs.org/en/download/), before I did this the build job would not start as the demands were not met.

As noted above I best guess is the most likely issue is the Node version, it must be Node 10 or higher.

Hope that helps, to make any more progress do a run the debug logging enabled and add it to this call

**Comment by mayeager22** [comment]
Thanks for the reply.  I will give it a try from an agent that has Node installed.

Thank you,

Mary Ann Yeager | Software Configuration Manager
Phidiax, LLC | www.phidiax.com<http://www.phidiax.com/>
Phone: 303-669-6944 | Main Office: 720-310-0161 | Email: MaryAnnYeager@phidiax.com<mailto:MaryAnnYeager@phidiax.com>

**Comment by mayeager22** [comment]
I have added a task to install version 10.x of node to the agent and added system.debug=true, but am still getting the same error.  Here is the log file (with collection url redacted).
[GenerateReleaseNotes_Error.txt](https://github.com/rfennell/AzurePipelines/files/5163983/GenerateReleaseNotes_Error.txt)

**Comment by rfennell** [comment]
OK, this might be a bit left field but could you try swapping the Release Notes task to use the V2 version (the drop down in the task parameters) as opposed to V3.

![image](https://user-images.githubusercontent.com/324718/92089439-10d62a80-edc6-11ea-8fc5-74e77a78e24c.png)

Also you will need to make sure the Agent Stage is set to allow access to the OAUTH token [see the docs] (https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes-V2--Node-based-Cross-Platform-Task) and the screenshot below

![image](https://user-images.githubusercontent.com/324718/92089525-2c413580-edc6-11ea-92fe-72df26e2aefd.png)

The reason to ask for this is that in V3 there is a change as to now the API token is obtained. In V3 it is got via the Azure DevOps Node SDK, whereas in V2 gets it via the OAUTH token (an environment variable).

If you re-run the release I hope you will get different results.

If it works you have a short term fix, which but I will need to add OAUTH support back into V3 to support whatever edge case you are experiencing  with your Azure DevOps Server 2019 instance. The strange thing is that I don't suffer this problem in my test VM 

On an unrelated notes I see you are using the legacy template format, this is not support with the V3 version of this task. V3 only supports handlebars templates. So moving to V2 (which supports both Handlebars and Legacy templates) so will allow the template you are using to work currently. However, once we are have fix for V3 you will need to move to [handlebar format]

**Comment by mayeager22** [comment]
I already had the OAuth setting on the agent job.  I have tried version 2 and am getting a similar error.

I will look for handlebars template examples, I just grabbed the template from your blog.

[GenerateReleaseNotes_ErrorV2.txt](https://github.com/rfennell/AzurePipelines/files/5169239/GenerateReleaseNotes_ErrorV2.txt)

**Comment by rfennell** [comment]
I doubt changing the template will have much effect as the problem is that the task cannot contact the Azure DevOps API.

I am at a bit of a loss on this one as by trying V2 and V3 you have tried two different ways get the access token. Also I have had no similar problems with other TFS/Azure DevOps Server installs 

I will have a bit more of a think.

**Comment by mayeager22** [comment]
If it helps, I am able to use the old PowerShell Generate Release Notes extension successfully.

**Comment by rfennell** [comment]
The old PowerShell task gets the API connection from the SDK

` $vstsEndpoint = Get-VstsEndpoint -Name SystemVssConnection -Require`

Whereas as the Node task in V2 uses a build variable (hence the need for OAUTH enabling)

`var accessToken = tl.getVariable("System.AccessToken");`

and in V3 we are back to an API call

`var accessToken = agentApi.getSystemAccessToken();`

These should all have the same effect i.e. have the same access as the build agent to the API.

The only thing I can think of that differs is that there is some handling in the PowerShell version for using a proxy

```
function Use-SystemWebProxy {
    <#
    .SYNOPSIS
        Make the subsequent web connections use the system web proxy.
    #>

    Write-Verbose "Using system web proxy."
    $proxy = [System.Net.WebRequest]::GetSystemWebProxy()
    $proxy.Credentials = [System.Net.CredentialCache]::DefaultNetworkCredentials
    [System.Net.WebRequest]::DefaultWebProxy = $proxy
}
```

This sounds a bit like [this unrelated project's issue](https://github.com/isomorphic-git/isomorphic-git/issues/202).

So I guess the question is 'is there an enterprises proxy between the server and the agent?'. If there is can you try without using it?

**Comment by mayeager22** [comment]
I do see "Using system web proxy" in the logs when I run the PowerShell version, so my assumption is that there is a proxy.

I am not sure how I would try without using the proxy, can you provide details?

**Comment by rfennell** [comment]
The proxy settings will be in the network settings on your agent machine. Your system admin should be able to tell you if you are using a proxy. Or you can check [something like this](https://helpdeskgeek.com/networking/internet-connection-problem-proxy-settings/)

We have seen proxies cause build systems no end of problems, so if they can ruled out it is a good idea.

**Comment by mayeager22** [comment]
I am trying to implement this at a client site, so my visibility for things is limited.  Asking the network admin about a proxy, his response is "Azure DevOps was installed on prem with outbound access only. I don't think there is a need for a proxy."

When I look at the build agent settings, I see "Automatically detect settings" as ON and all others are OFF.

**Comment by mayeager22** [comment]
Is there any specific port this extension is using that might need to be allowed through the firewall?

**Comment by rfennell** [comment]
All TFS/Azure DevOps Server is by default either on port 8080 or 443 (if you are using SSL as I would always recommend)

**Comment by mayeager22** [comment]
We are using SSL, but it is port 8443 instead.  Would that cause an issue with this extension?

**Comment by rfennell** [comment]
No should not be an issue, the key 8s if it goes via a proxy, via 8443, that could be the problem, at least something to rule out

**Comment by mayeager22** [comment]
I do not believe it is going through a proxy.

**Comment by rfennell** [comment]
I am not sure what to suggest as a cannot replicate your problem on my test system.

I guess we are into extremes like
- Un-install my extension and then reinstall it - I have seen corrupt installed on premises 
- try a different build agent - in case the issue is the local PC/VM
- try installing the build agent on the server to see if you get the same problem. If it works on the server and not on a client it suggests some strange network issue
- You said you are using 8443 is there anything strange about the cert?
- Can you try registering the agent on port 8080 see if that changes anything

**Comment by mayeager22** [comment]
- I have already uninstalled and reinstalled the extension.
- I tried a different build agent when we thought the issue was the version of Node.
- Using a new build agent created directly on the application server gives the same error.
- The only thing I can think of that might be considered "strange" is that we need to use the FQDN of the server (https://servername.inside.companyname.org:8443/tfs) for the URL
- There is a redirect on port 8080, so I am unable to register the agent using that port.

**Comment by rfennell** [comment]
It is certainly strange. 

You mention the FQDN, maybe you could try http://locahost:8080/tfs when the agent is on the actual TFS server

**Comment by mayeager22** [comment]
Using http://locatlhost:8080/tfs also does not allow me to register the agent.

However, I have a test Azure DevOps Server 2019 installation I used to test an upgrade for this customer.  If I disable the redirect, I am able to register the agent using port 8080 and the task runs without the error.

So, it does have something to do with redirecting from 8080 to 8443.

**Comment by rfennell** [comment]
Strange, how are you doing the redirect ?

The way I would expect is with in IIS bind both port 8080 (HTTP) and 8443 (HTTPS) to the TFS virtual application. Then in Azure DevOps Server admin console set the public IP address to be the https://servername.inside.companyname.org:8443/tfs address (so the server all refers to it self as this address)

If you want no access to the server 8080 across the network, then I would block it on the server firewall.

I have used this setup for many TFS Azure DevOps Server instances over the years without a problem - does yours differ?

**Comment by mayeager22** [comment]
That is exactly how I have it set up.

**Comment by rfennell** [comment]
Well I am really stumped then. It must be something to do with the server setup but I can't think what.

**Comment by rfennell** [comment]
After another google does this help https://stackoverflow.com/questions/56855546/azure-devops-server-pipeline-build-fails-when-using-self-signed-ssl-certificate

It seems there are some issue with Node reading cert details in some circumstances. It is not my task that is the issue but the underlying use of Node.

**Comment by mayeager22** [comment]
I added the PowerShell task suggested in that link, but now I am getting an additional error before the "local issuer certificate error":

2020-09-22T19:48:33.6987184Z (node:14640) Warning: Ignoring extra certs from `D:\Builds\Agent1\TrustedRootCAs.pem`, load failed: error:0B07C065:x509 certificate routines:X509_STORE_add_cert:cert already in hash table

**Comment by rfennell** [comment]
The only other question I would ask is 'is the SSL cert you are using self signed or one off a recognised root certificate?'. 

Either way, I think this has reached the point I cannot help anymore. We have seen the task work on your second server and I have it working for many other users. So it really looks like an issue with your server or the cert.

So I think the next step will be to be for you to raise a support ticket with Microsoft.

**Comment by mayeager22** [comment]
Thank you for taking the time to review this with me.  I do have the PowerShell version of the task working, so I will probably just stick with that.

**Comment by rfennell** [comment]
I do have one final option. 

I could do you a build that allows you to inject a PAT token as a parameter (to bypass the use of the agent OAUTH token). The inner task functions already include code to allow this authentication model for use when running in a test harness, but at present it is not exposed within Azure DevOps task usage. But it is not hard to add. 

If it works for you then I can then publish and update the public task with this option as an edge case authentication fall back.

I obviously don't want to waste any more of your time if you are happy with the PowerShell version. But this option is there if it helps

**Comment by mayeager22** [comment]
The PowerShell version is working fine for me.  But I would be happy to test out the PAT option in case others run into this issue.

**Comment by rfennell** [comment]
I have just published 3.15.3 to the marketplace. This has the new feature to allow you to pass a PAT in via the property `overridePat` (in advanced section using UI based tools as opposed to YAML).

Hopefully this will allow you to make progress by passing in a PAT that has rights on the server.

However, my concern is that as your initial log shows the failure well before any of my code is loaded, as Node starts up. So even with this change it won't help you.

Anyway, it is a useful addition for edge cases and testing, so let me know how it goes.

**Comment by mayeager22** [comment]
I don't see this version in the marketplace yet (latest I see is 13.14.4).

**Comment by rfennell** [comment]
The Marketplace management tool says it is published - strange

You can also download it from where it is [built here](https://dev.azure.com/richardfennell/GitHub/_build/results?buildId=8298&view=results)

And I have attached it to this issue
[richardfennellBM.BM-VSTS-XplatGenerateReleaseNotes-3.15.3.zip](https://github.com/rfennell/AzurePipelines/files/5277316/richardfennellBM.BM-VSTS-XplatGenerateReleaseNotes-3.15.3.zip)

**Comment by mayeager22** [comment]
The extension did eventually update automatically.  I have tried version 3.* using a PAT and am getting a very similar message:

2020-09-28T15:17:55.7181058Z Creating the credential handler using override PAT (suitable for local testing or if the OAUTH token cannot be used)
2020-09-28T15:17:55.7720564Z (node:8728) UnhandledPromiseRejectionWarning: Error: unable to get local issuer certificate
2020-09-28T15:17:55.7721286Z     at TLSSocket.onConnectSecure (_tls_wrap.js:1049:34)
2020-09-28T15:17:55.7721652Z     at TLSSocket.emit (events.js:182:13)
2020-09-28T15:17:55.7722035Z     at TLSSocket._finishInit (_tls_wrap.js:631:8)
2020-09-28T15:17:55.7722512Z (node:8728) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
2020-09-28T15:17:55.7722975Z (node:8728) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

So, I think I will just continue to use the PowerShell version.

**Comment by rfennell** [comment]
Yes, I think that is your best option

Hope you don't suffer problems with other Node based tasks that need to use the API

Thanks for persisting in try to resolve this, sorry that I can't progress this any more.

Richard

**Comment by Eneuman** [comment]
I think you can get around the Node.js issue by adding the variable "NODE_TLS_REJECT_UNAUTHORIZED" with value "0" to your release pipeline but it has security implications.

**Comment by rfennell** [comment]
Thanks @Eneuman for that idea. The easiest way to set that is to add an Azure DevOps variable 

`NODE_TLS_REJECT_UNAUTHORIZED=0`

All the Azure DevOps variables are set as environment variables in the agent, so this should have the desired effect.

@mayeager22 does this help?

**Comment by mayeager22** [comment]
@rfennell Yes, adding this variable to the release pipeline allows the task to run successfully.

**Comment by rfennell** [comment]
Great, I think we can close this issue then - thanks again @Eneuman  for the solution

**Comment by Eneuman** [comment]
I think the main problem is that nodejs (that is used by the agent) is not using the windows certificate store, and the certificate issued to the server is self signed or from a in-house CA root.

You are going to need to add your Azure DevOps server certificate as a EXTRA CERTIFICATE to node js if you want to remove the workaround.

**Comment by niklev** [comment]
Is there any update on this? Having the same issue on Linux hosted agents behind SSL proxy.

**Comment by niklev** [comment]
To specify certificate of ssl proxy you need to add a variable on the task as such (must better solution than turning off TLS verification) - in the example the corporate proxy certificate has been added to agent's ca-certificates.crt :

jobs:
        ...
        variables: 
          NODE_EXTRA_CA_CERTS: /etc/ssl/certs/ca-certificates.crt
          ...
          - task: GitHubRelease@1
                    displayName: 'Create GitHub Release'

**Comment by NecatiMeral** [comment]
Sorry for reviving this thread. I've experienced the same issue but I solved the issue on the agent itself instead of adding this variable everywhere.

1. `ssh` into your agent and create the `NODE_EXTRA_CA_CERTS` variable for all users
2. Open the configured `azagent` directory and append the variable to the [`env.sh`](https://github.com/microsoft/azure-pipelines-agent/blob/master/src/Misc/layoutroot/env.sh) file and run the following commands:

```bash
./env.sh
sudo ./svc.sh stop
sudo ./svc.sh start
```

This will make the agent respect the environment variable globally and should keep your pipelines clear.

Source: https://docs.microsoft.com/de-de/azure/devops/pipelines/agents/v2-linux?view=azure-devops#service-update-environment-variables
